noImplicitReturns misidentifies generator return as function return when strictNullChecks is false

**TypeScript Version:** 2.6.1 -  2.7.0-dev.20171128

**Code:**
```ts
// index.ts
function* testGenerator() {
    if (Math.random() > 0.5) {
        return;
    }
    yield 'hello';
}

/* tsconfig.json
{
    "files": ["index.ts"],
    "compilerOptions": {
        "lib": ["esnext"], "target": "ESNEXT", "module": "es2015",
        "strictNullChecks": true,
        "noImplicitReturns": true
    }
}
*/
```

**Expected behavior:**

No errors due to the generator's early `return`, regardless of whether `strictNullChecks` is true or false.

**Actual behavior:**

* No errors when `strictNullChecks` is _true_
* The following error when `strictNullChecks` is _false_:
  ```index.ts(3,9): error TS7030: Not all code paths return a value.```

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -22054,17 +22054,16 @@ namespace ts {
             if (func) {
                 const signature = getSignatureFromDeclaration(func);
                 const returnType = getReturnTypeOfSignature(signature);
+                const functionFlags = getFunctionFlags(func);
+                if (functionFlags & FunctionFlags.Generator) { // AsyncGenerator function or Generator function
+                    // A generator does not need its return expressions checked against its return type.
+                    // Instead, the yield expressions are checked against the element type.
+                    // TODO: Check return expressions of generators when return type tracking is added
+                    // for generators.
+                    return;
+                }
                 if (strictNullChecks || node.expression || returnType.flags & TypeFlags.Never) {
                     const exprType = node.expression ? checkExpressionCached(node.expression) : undefinedType;
-                    const functionFlags = getFunctionFlags(func);
-                    if (functionFlags & FunctionFlags.Generator) { // AsyncGenerator function or Generator function
-                        // A generator does not need its return expressions checked against its return type.
-                        // Instead, the yield expressions are checked against the element type.
-                        // TODO: Check return expressions of generators when return type tracking is added
-                        // for generators.
-                        return;
-                    }
-
                     if (func.kind === SyntaxKind.SetAccessor) {
                         if (node.expression) {
                             error(node, Diagnostics.Setters_cannot_return_a_value);

--- a/tests/baselines/reference/generatorNoImplicitReturns.js
+++ b/tests/baselines/reference/generatorNoImplicitReturns.js
@@ -1,0 +1,17 @@
+//// [generatorNoImplicitReturns.ts]
+ 
+function* testGenerator () { 
+  if (Math.random() > 0.5) { 
+      return; 
+  } 
+  yield 'hello'; 
+}
+
+
+//// [generatorNoImplicitReturns.js]
+function* testGenerator() {
+    if (Math.random() > 0.5) {
+        return;
+    }
+    yield 'hello';
+}

--- a/tests/baselines/reference/generatorNoImplicitReturns.symbols
+++ b/tests/baselines/reference/generatorNoImplicitReturns.symbols
@@ -1,0 +1,15 @@
+=== tests/cases/conformance/es6/yieldExpressions/generatorNoImplicitReturns.ts ===
+ 
+function* testGenerator () { 
+>testGenerator : Symbol(testGenerator, Decl(generatorNoImplicitReturns.ts, 0, 0))
+
+  if (Math.random() > 0.5) { 
+>Math.random : Symbol(Math.random, Decl(lib.es5.d.ts, --, --))
+>Math : Symbol(Math, Decl(lib.es5.d.ts, --, --), Decl(lib.es5.d.ts, --, --), Decl(lib.es2015.symbol.wellknown.d.ts, --, --), Decl(lib.es2015.core.d.ts, --, --))
+>random : Symbol(Math.random, Decl(lib.es5.d.ts, --, --))
+
+      return; 
+  } 
+  yield 'hello'; 
+}
+

--- a/tests/baselines/reference/generatorNoImplicitReturns.types
+++ b/tests/baselines/reference/generatorNoImplicitReturns.types
@@ -1,0 +1,20 @@
+=== tests/cases/conformance/es6/yieldExpressions/generatorNoImplicitReturns.ts ===
+ 
+function* testGenerator () { 
+>testGenerator : () => IterableIterator<string>
+
+  if (Math.random() > 0.5) { 
+>Math.random() > 0.5 : boolean
+>Math.random() : number
+>Math.random : () => number
+>Math : Math
+>random : () => number
+>0.5 : 0.5
+
+      return; 
+  } 
+  yield 'hello'; 
+>yield 'hello' : any
+>'hello' : "hello"
+}
+

--- a/tests/cases/conformance/es6/yieldExpressions/generatorNoImplicitReturns.ts
+++ b/tests/cases/conformance/es6/yieldExpressions/generatorNoImplicitReturns.ts
@@ -1,0 +1,10 @@
+// @target: esnext 
+// @noImplicitReturns: true 
+// @strictNullChecks: false 
+ 
+function* testGenerator () { 
+  if (Math.random() > 0.5) { 
+      return; 
+  } 
+  yield 'hello'; 
+}

